### PR TITLE
Common module base class and module documentation

### PR DESCRIPTION
**Pull request by bkueng**
We currently have these problems:
- module startup & shutdown are susceptible to race conditions and failure cases are generally not handled well, it's easy to make mistakes (#1644)
- modules documentation is missing completely on the dev guide ( @hamishwillee )
- the help function of some modules is not useful, incomplete or missing altogether. I sometimes have to go through the code to see how a command works.

This PR proposes a solution for these. Although the documentation & common base class are independent problems (and my solution), it makes sense to change them in one go.

### Module base class
Design goals:
- should work with threads and work queues
- integrate well with current structure to minimize the converting effort
- constructor & destructors should be called from within the new task (so that orb_subscribe etc work)
- thread-safe
- provide only the basic functionality (We already have the Block classes for param handling for example)
- possibility to wait until a task is started
- the `px4_task_spawn_cmd` and `px4_getopt` should be handled in the derived class, to provide customization

I used a template class so that it can have a static instance that is different for each derived module.

Limitations:
- it currently does not work with multi-instance modules such as mavlink, or some drivers (eg. mpu6000), but it could be extended for that.
- `errx` and `err` do not work, since they can leave the lock in a locked state. Since we wanted to remove these anyway, this is a good opportunity (we can get rid of `warnx` too while at it).

### Module documentation & usage
My idea is to have something similar to a `man` page that can be used on the console, as an online documentation, and in the source. It should be in the source, so that it's easy to update and in sync with the module. It's a document-once-use-everywhere approach to minimize maintainance efforts.

The module documentation looks like this (to make it readable in the source, I used C++11 raw string literals):
```
	PRINT_MODULE_DESCRIPTION(
		R"DESCR_STR(
System logger which logs a (configurable) set of uORB topics and system printf messages
(warnings and errors) to ULog files. These can be used for system and flight performance evaluation,
tuning, replay and crash analysis.
[...]
Typical usage to start logging immediately:
$ logger start -e -t
)DESCR_STR");
```
The intention is that it includes:
- provided functionality of the module
- high-level implementation overview
- examples how to use the CLI interface

These strings can then extracted via script and turned into a html page (we could use markdown in the description and add sections). I haven't done that yet.

There are some limitations, for instance, no images, so it will still make sense to have a separate documentation page for a few of the modules.

On NuttX the module description is removed, to reduce Flash usage.

Similarly the usage help strings: they should be standardized, so that we can:
- easily use them on the console
- extract them and show them online
- maybe do fuzzy testing later on, allowing to test less common code paths in an automated way

This is the usage output for the `pwm` command (probably one of the more complex ones):
```
Usage: pwm <command> [arguments...]
 Commands:

   arm           Arm output

   disarm        Disarm output

   info          Print current configuration of all channels

   forcefail     Force Failsafe mode
     on|off      Turn on or off

   terminatefail Force Termination Failsafe mode
     on|off      Turn on or off

   rate          Configure PWM rates
     -r <val>    PWM Rate in Hz (0 = Oneshot, otherwise 50 to 400Hz)

   oneshot       Configure Oneshot125 (rate is set to 0)

   failsafe      Set Failsafe PWM value

   disarmed      Set Disarmed PWM value

   min           Set Minimum PWM value

   max           Set Maximum PWM value

   test          Set Output to a specific value until 'q' or 'c' or 'ctrl-c' pressed

   steps         Run 5 steps from 0 to 100%


 The commands 'failsafe', 'disarmed', 'min', 'max' and 'test' require a PWM value:
     -p <val>    PWM value (eg. 1100)

 The commands 'rate', 'oneshot', 'failsafe', 'disarmed', 'min', 'max', 'test' and 'steps' additionally require to specify the channels with one of the following commands:
     [-c <val>]  select channels in the form: 1234 (1 digit per channel, 1=first)
     [-m <val>]  Select channels via bitmask (eg. 0xF, 3)
                 default: 0
     [-g <val>]  Select channels by group (eg. 0, 1, 2. use 'pwm info' to show groups)
                 default: 0
     [-a]        Select all channels

 These parameters apply to all commands:
     [-d <val>]  Select PWM output device
                 values: <file:dev>, default: /dev/pwm_output0
     [-v]        Verbose output
     [-e]        Exit with 1 instead of 0 on error
```

I already converted some modules & commands to test that it works (logger, send_events and fmu, which can run as task or on the work queue).
Others should work too, I looked at commander, dataman, navio_sysfs_pwm, mc_pos_controller, navigator, ekf2, LPE, fw_pos_control, fw_att_control, and land_detector.

Any thoughts?

### CODE REVIEW

**Comment by TSC21** [comment]
Super useful!!!

File: src/modules/logger/logger.h [hunk 1]
```diff
@@ -80,14 +81,32 @@ struct LoggerSubscription {
...
 	Logger(LogWriter::Backend backend, size_t buffer_size, uint32_t log_interval, const char *poll_topic_name,
 	       bool log_on_start, bool log_until_shutdown, bool log_name_timestamp, unsigned int queue_size);
 
 	~Logger();
 
+	/** @see ModuleBase */
+	static int task_spawn(int argc, char *argv[]);
+
+	/** @see ModuleBase */
+	static Logger *instanciate(int argc, char *argv[]);
```
Comment (dagar) [inline src/modules/logger/logger.h]: instanciate -> instantiate

**Comment by dagar** [comment]
This is great! 

Somewhat unrelated, but what about general guidance for when to use threads vs a work queue (and which one)?

**Comment by LorenzMeier** [comment]
Nice! And really, really needed.

File: src/modules/mavlink/mavlink_main.cpp [hunk 2]
```diff
@@ -2699,22 +2700,63 @@ Mavlink::set_boot_complete()
 
 static void usage()
 {
-	PX4_INFO("usage: mavlink {start|status|stream|stop-all|boot_complete}");
-	PX4_INFO("    [-d device]");
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
```
Comment (LorenzMeier) [inline src/modules/mavlink/mavlink_main.cpp]: The documentation here is 1K flash. Just to keep that in perspective.

File: src/modules/logger/logger.cpp [hunk 3]
```diff
@@ -110,122 +107,100 @@ int logger_main(int argc, char *argv[])
...
-		 "\t-p\tPoll on a topic instead of running with fixed rate (Log rate and topic interval are ignored if this is set)\n"
-		 "\t-b\tLog buffer size in KiB, default is 12\n"
-		 "\t-e\tEnable logging right after start until disarm (otherwise only when armed)\n"
-		 "\t-f\tLog until shutdown (implies -e)\n"
-		 "\t-t\tUse date/time for naming log directories and files\n"
-		 "\t-m\tMode: one of 'file', 'mavlink', 'all' (default=all)\n"
-		 "\t-q\tuORB queue size for mavlink mode");
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+System logger which logs a (configurable) set of uORB topics and system printf messages
```
Comment (LorenzMeier) [inline src/modules/logger/logger.cpp]: 1K flash

File: src/drivers/px4fmu/fmu.cpp [hunk 4]
```diff
@@ -3470,81 +3462,153 @@ fmu_main(int argc, char *argv[])
...
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+This module is responsible for driving the output and reading the input pins for boards without a separate IO chip
+(eg. Pixracer). On boards with an IO chip, the px4io driver is used.
 
-		} else {
-			sensor_reset(0);
-			PX4_INFO("reseted default time");
-		}
+It listens on the actuator_controls topics, does the mixing and writes the PWM outputs.
```
Comment (LorenzMeier) [inline src/drivers/px4fmu/fmu.cpp]: 1K flash 8)

**Comment by hamishwillee** [comment]
Looks very cool. Certainly makes sense to include comments in source for multiple use. 

> There are some limitations, for instance, no images, so it will still make sense to have a separate documentation page for a few of the modules.

It would be good to work around no images. For the number of cases we're talking, would it be OK if images were specified in markdown format (as URLs to Github say) and rendered in console. Or could we have a smart console printer that converts markdown images to a more readable format for console?

File: src/modules/logger/logger.h (hunk 1 above)
Comment (bkueng) [inline src/modules/logger/logger.h]: Fixed

File: src/modules/mavlink/mavlink_main.cpp (hunk 2 above)
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp]: It's removed on NuttX: https://github.com/PX4/Firmware/pull/7145/commits/62d4cb63e0c443851164a85e1eeab9d5eb69936e#diff-1d74d6cd3d156b55d40085f178a49cfeR356

**Comment by bkueng** [comment]
Thanks all for the feedback!

> It would be good to work around no images. For the number of cases we're talking, would it be OK if images were specified in markdown format (as URLs to Github say) and rendered in console. Or could we have a smart console printer that converts markdown images to a more readable format for console?

I think we should go for markdown then. If it looks really bad on the console, we can still do some conversion.

**Comment by bkueng** [comment]
I switched to Markdown & added a script to extract the documentation. First results are here:
https://github.com/bkueng/Firmware/blob/modules_doc/Tools/modules.md. We can still adjust the layout, eg. use one page per category.

The script also does some consistency checks, to test if the documented commands are in sync with the code: it checks the documented options against the ones in `px4_getopt`, and checks the commands against occurrences of `strcmp(... argv, "<command>")`. This is only approximative (nevertheless I already found 2 things), and I kept it conservative, such that it should not produce false positives. Currently it only outputs warnings, but does not let the script fail. @dagar if it works well we could add this to CI.

I suggest we merge this after the release, and then we can do the conversion in stages. Any help is appreciated. (Daniel already volunteered for the FW controllers, thanks for that)

@hamishwillee I updated the airframes script to optionally specify the image path: `./px_process_airframes.py -m -i ../image/path`

**Comment by dagar** [comment]
Trying to collect all metadata generation here.
https://github.com/PX4/Firmware/blob/master/Makefile#L251

**Comment by bkueng** [comment]
Ok I'll add it there. Will this run for PR's as well?

**Comment by dagar** [comment]
Yes, although we still need to find somewhere to run this that can automatically push to the gitbook.

**Comment by hamishwillee** [comment]
@bkueng Looks great! The docs with descriptions/examples are much more useful to me than those without. We should add at least a paragraph description for all cases.

> We can still adjust the layout, e.g. use one page per category. 

One page per category makes sense if the page TOC needs to be scrolled to see all the content,

> I updated the airframes script to optionally specify the image path

Great. Will scripts like that one get documented using this tool too?

**Comment by bkueng** [comment]
All the system commands are documented now.

**Comment by hamishwillee** [comment]
So this just requires review/approval? Is getting the script run and the docs integrated into the dev guide a separate PR?

**Comment by bkueng** [comment]
I planned to merge this after the release, as it contains changes to `fmu`, a very central and important driver. Meanwhile I'll continue a bit here:
- adding a Makefile target for CI
- try out some different layouts for the dev-guide to see what works best

File: src/modules/mavlink/mavlink_main.cpp [hunk 5]
```diff
@@ -2699,22 +2700,70 @@ Mavlink::set_boot_complete()
...
+	PRINT_MODULE_USAGE_COMMAND_DESCR("stream", "Configure the sending rate of a stream for a running instance");
+#ifdef __PX4_POSIX
+	PRINT_MODULE_USAGE_PARAM_INT('u', 0, 0, 65536, "Select Mavlink instance via local Network Port", true);
+#endif
+	PRINT_MODULE_USAGE_PARAM_STRING('d', nullptr, "<file:dev>", "Select Mavlink instance via Serial Device", true);
+	PRINT_MODULE_USAGE_PARAM_STRING('s', nullptr, nullptr, "Mavlink stream to configure", false);
+	PRINT_MODULE_USAGE_PARAM_FLOAT('r', 0.f, 0.f, 2000.f, "Rate in Hz (0 = turn off)", false);
+
+	PRINT_MODULE_USAGE_COMMAND_DESCR("boot_complete",
+					 "Enable sending of messages. (Must be) called as last step in startup script.");
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_main.cpp]: Aha, that's what it does :)

File: src/modules/mavlink/mavlink_main.cpp [hunk 6]
```diff
@@ -2699,22 +2700,70 @@ Mavlink::set_boot_complete()
...
+**Careful**: some of the data is accessed and modified from both threads, so when changing code or extend the
+functionality, this needs to be take into account, in order to avoid race conditions and corrupt data.
+
+### Examples
+Start mavlink on ttyS1 serial with baudrate 921600 and maximum sending rate of 80kB/s:
+$ mavlink start -d /dev/ttyS1 -b 921600 -m onboard -r 80000
+
+Start mavlink on UDP port 14556 and enable the HIGHRES_IMU message with 50Hz:
+$ mavlink start -u 14556 -r 1000000
+$ mavlink stream -u 14556 -s HIGHRES_IMU -r 50
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_main.cpp]: Awesome docs!

File: src/platforms/px4_module.h [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+		if (strcmp(argv[1], "stop") == 0) {
+			return stop_command();
+		}
+
+		if (strcmp(argv[1], "status") == 0) {
+			return status_command();
+		}
+
+		lock_module(); // we better lock here, as the method could access _object
+		int ret = T::custom_command(argc - 1, argv + 1);
```
Comment (julianoes) [inline src/platforms/px4_module.h]: Cool!

File: src/platforms/px4_module.h [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+		if (strcmp(argv[1], "start") == 0) {
+			// we pass the 'start' argument too, because later on px4_getopt() will ignore the first argument
+			return start_command_base(argc - 1, argv + 1);
+		}
+
+		if (strcmp(argv[1], "stop") == 0) {
+			return stop_command();
+		}
+
+		if (strcmp(argv[1], "status") == 0) {
```
Comment (julianoes) [inline src/platforms/px4_module.h]: Good, so `status` it is, and `info` should go away.

File: src/platforms/px4_module.h [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+	 * @param argc start argument(s)
+	 * @param argv start argument(s)
+	 */
+	static void run_trampoline(int argc, char *argv[])
+	{
+
+#ifdef __PX4_NUTTX
+		// on NuttX task_create() adds the task name as first argument
+		argc -= 1;
+		argv += 1;
```
Comment (julianoes) [inline src/platforms/px4_module.h]: Is this still needed?

File: src/platforms/px4_module.h [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+
+	// there will be one instance for each template type
+	static volatile T *_object; ///< instance if the module is running
+	static int _task_id;        ///< task handle: -1 = invalid, otherwise task is assumed to be running
+
+	static constexpr const int task_id_is_work_queue = -2; ///< special value if task runs on the work queue
+
+private:
+
+	volatile bool _task_should_exit = false;
```
Comment (julianoes) [inline src/platforms/px4_module.h]: I thought this should be atomic instead of volatile?

File: src/platforms/px4_module.h [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+ * @param is_optional true if this parameter is optional
+ */
+__EXPORT void PRINT_MODULE_USAGE_PARAM_INT(char option_char, int default_val, int min_val, int max_val,
+		const char *description, bool is_optional);
+
+/**
+ * print a float parameter
+ * @see PRINT_MODULE_USAGE_PARAM_INT()
+ */
+__EXPORT void PRINT_MODULE_USAGE_PARAM_FLOAT(char option_char, float default_val, float min_val, float max_val,
```
Comment (julianoes) [inline src/platforms/px4_module.h]: fancy!

File: src/systemcmds/pwm/pwm.cpp [hunk 12]
```diff
@@ -66,59 +67,90 @@
...
+					true);
+	PRINT_MODULE_USAGE_PARAM_INT('m', 0, 0, 4096, "Select channels via bitmask (eg. 0xF, 3)", true);
+	PRINT_MODULE_USAGE_PARAM_INT('g', 0, 0, 10, "Select channels by group (eg. 0, 1, 2. use 'pwm info' to show groups)",
+				     true);
+	PRINT_MODULE_USAGE_PARAM_FLAG('a', "Select all channels", true);
+
+	PRINT_MODULE_USAGE_PARAM_COMMENT("These parameters apply to all commands:");
+	PRINT_MODULE_USAGE_PARAM_STRING('d', "/dev/pwm_output0", "<file:dev>", "Select PWM output device", true);
+	PRINT_MODULE_USAGE_PARAM_FLAG('v', "Verbose output", true);
+	PRINT_MODULE_USAGE_PARAM_FLAG('e', "Exit with 1 instead of 0 on error", true);
```
Comment (julianoes) [inline src/systemcmds/pwm/pwm.cpp]: Wow, this is so much cleaner!

File: src/drivers/px4fmu/fmu.cpp [hunk 13]
```diff
@@ -189,9 +235,8 @@ class PX4FMU : public device::CDev
 	unsigned	_pwm_alt_rate;
 	uint32_t	_pwm_alt_rate_channels;
 	unsigned	_current_update_rate;
-	int 		_task;
 	bool 		_run_as_task;
-	struct work_s	_work;
+	static struct work_s	_work;
```
Comment (bkueng) [inline src/drivers/px4fmu/fmu.cpp]: There's a dependency problem: we want to instantiate the object within the work queue, but we need the `work_s` item to kick off the work queue before that.

File: src/drivers/px4fmu/fmu.cpp [hunk 14]
```diff
@@ -3479,81 +3482,165 @@ fmu_main(int argc, char *argv[])
...
+	switch (_mode) {
+	case MODE_1PWM: mode_str = "pwm1"; break;
+
+	case MODE_2PWM: mode_str = "pwm2"; break;
+
+	case MODE_3PWM: mode_str = "pwm3"; break;
+
+	case MODE_4PWM: mode_str = "pwm4"; break;
+
+	case MODE_2PWM2CAP: mode_str = "pwm2cap1"; break;
```
Comment (bkueng) [inline src/drivers/px4fmu/fmu.cpp]: Oh, this is wrong

File: src/platforms/px4_module.h (hunk 9 above)
Comment (bkueng) [inline src/platforms/px4_module.h]: Yes, I wasn't expecting that either

File: src/platforms/px4_module.h (hunk 10 above)
Comment (bkueng) [inline src/platforms/px4_module.h]: Ideally yes, but we don't have an atomic API yet...

**Comment by bkueng** [comment]
Julian, I addressed you comments, thanks for the review

**Comment by dagar** [comment]
A couple questions.

1. Could we stack allocate the object instead? This would provide better visibility into memory usage on NuttX where I don't think we can (easily?) see heap usage per process. It also keeps things simple and as a general rule we wouldn't have any dynamic allocation in most modules.

2. Could the base have a default task spawn? There could be a thread and work queue module base with relevant arguments (priority and thread stack size or queue and update rate) passed to the base constructor, or defined at the cmake level. 

3. Can the helper functions (print_usage, custom_command) be virtual with empty default implementations.

Overall what I'm getting at is most modules having almost no boiler plate at all. At a minimum just inherit from the appropriate base, implement update(), and go.

**Comment by bkueng** [comment]
> Could we stack allocate the object instead? This would provide better visibility into memory usage on NuttX where I don't think we can (easily?) see heap usage per process. It also keeps things simple and as a general rule we wouldn't have any dynamic allocation in most modules. I like how lpe is structured - https://github.com/PX4/Firmware/blob/master/src/modules/local_position_estimator/local_position_estimator_main.cpp#L161

It might give you a rough estimate. But: modules can and will still dynamically allocate memory (I'd say about half of the modules). That would still be hidden. In general I found that memory analysis is best done on POSIX, yielding more accurate and fine-grained results (just knowing that a module allocates a huge block does not help much, and you can also see that from starting/stopping the module and look at the free memory difference). It actually becomes more difficult when keeping everything on the stack, since you then need to look at the separate stack frames.
The results on POSIX are relevant on NuttX too.
At the same time I'm also in favor of reducing the number of dynamic allocations, so if you come up with a nice solution I could go with that as well.

> Can the helper functions (print_usage, custom_command) be virtual with default implementations for start, stop, status (running or not)? This alone would be enough for many modules including all the controllers.

I tried to do that as much as possible (for example the status is optional), but it's not possible for the static methods. If you have a better way, I'll gladly go with that.
An important point to me is to reduce boiler-plate code as well, but while keeping it generic enough to work with the special cases as well. And I found that these are actually not uncommon, only a minority of the modules just needs start/stop/status.

> Could the base have a default task spawn? There could be a thread and work queue module base with relevant arguments (priority and thread stack size or queue and update rate) passed to the base constructor, or defined at the cmake level.

Same problem with the static, and I wanted to avoid adding another layer, so that the code can be greped for `px4_task_spawn_cmd`. With separate base classes, how would you solve the fmu-case that can run on the work queue and as task?

**Comment by dagar** [comment]
Do we know why the fmu needs to support both? Could we accomplish the same thing by parameterizing the work queue scheduling or task priority instead?

**Comment by bkueng** [comment]
It's a tradeoff between latency (the task directly polls on the topic) and RAM (the task needs it's own stack & task overhead). So no, I don't think parameterizing works.

**Comment by LorenzMeier** [comment]
@bkueng Can you rebase and fix build? Then we can bring this in.

**Comment by bkueng** [comment]
Rebased & CI passes, this is good to go

**Comment by bkueng** [comment]
Closing, as #7616 got merged & github fixed the access.

**Pull request closed without merging**